controllers/pipelines/provider_controller.go

```diff
@@ -0,0 +1,230 @@
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	AppLabel               = "app"
+	ResourceHashAnnotation = "resource-hash"
+)
+
+type ProviderReconciler struct {
+	ResourceReconciler[*pipelinesv1.Provider]
+	Scheme *runtime.Scheme
+}
+
+func NewProviderReconciler(ec K8sExecutionContext, config config.KfpControllerConfigSpec) *ProviderReconciler {
+	return &ProviderReconciler{
+		ResourceReconciler: ResourceReconciler[*pipelinesv1.Provider]{
+			EC:     ec,
+			Config: config,
+		},
+		Scheme: ec.Scheme,
+	}
+}
+
+//+kubebuilder:rbac:groups=pipelines.kubeflow.org,resources=providers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=pipelines.kubeflow.org,resources=providers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=pipelines.kubeflow.org,resources=providers/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	startTime := time.Now()
+	logger.Info("reconciliation started", "request", req)
+
+	var provider = &pipelinesv1.Provider{}
+	if err := r.EC.Client.NonCached.Get(ctx, req.NamespacedName, provider); err != nil {
+		logger.Error(err, "unable to get provider")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger.Info("found provider", "resource", provider)
+
+	desiredDeployment, err := constructDeployment(provider, *r.Config.DeepCopy())
+	if err != nil {
+		logger.Error(err, "unable to construct provider deployment")
+		return ctrl.Result{}, err
+	}
+
+	if err := ctrl.SetControllerReference(provider, desiredDeployment, r.Scheme); err != nil {
+		logger.Error(err, "unable to set controller reference on deployment")
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("constructed desired provider deployment", "deployment", desiredDeployment)
+
+	existingDeployment, err := r.getDeployment(ctx, *provider)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logger.Error(err, "unable to get existing deployment")
+		return ctrl.Result{}, err
+	}
+
+	if existingDeployment != nil {
+		logger.Info("found existing provider deployment", "deployment", existingDeployment)
+
+		if err := setResourceHashAnnotation(existingDeployment); err != nil {
+			logger.Error(err, "unable to set resource hash annotation on existing deployment")
+			return ctrl.Result{}, err
+		}
+
+		if deploymentIsOutOfSync(existingDeployment, desiredDeployment) {
+			logger.Info("resource hash mismatch, updating deployment")
+			existingDeployment = syncDeployment(existingDeployment, desiredDeployment)
+			if err = r.EC.Client.Update(ctx, existingDeployment); err != nil {
+				logger.Error(err, "unable to update provider service deployment", "deployment", desiredDeployment)
```

Comment:
```suggestion
			desiredDeployment = syncDeployment(existingDeployment, desiredDeployment)
			if err = r.EC.Client.Update(ctx, desiredDeployment); err != nil {
				logger.Error(err, "unable to update provider service deployment", "deployment", desiredDeployment)
```
would it make more sense to replace value of `desiredDeployment` here?

Reply:
Well..... depends on how you conceptualise it I guess. The idea was we take the existing deployment in the cluster and update it based on the desired deployment. So the thing being "synced" is the existing deployment.

Kind of like how Argo CD Syncs an "existing" resource based on the "desired" version of that resource on your git repo.

Reply:
I would say it reads better my way but am not precious.

Reply:
Alrighty I'll just keep it as is then ✅